ecs-init/docker/dependencies.go

```diff
@@ -32,22 +37,48 @@ type dockerclient interface {
 	StartContainer(id string, hostConfig *godocker.HostConfig) error
 	WaitContainer(id string) (int, error)
 	StopContainer(id string, timeout uint) error
+	Ping() error
 }
 
 type _dockerclient struct {
-	docker *godocker.Client
+	docker dockerclient
 }
 
-func newDockerClient() (*_dockerclient, error) {
+type dockerClientFactory interface {
+	NewVersionedClient(endpoint string, apiVersionString string) (dockerclient, error)
+}
+
+type godockerClientFactory struct{}
+
+func (client godockerClientFactory) NewVersionedClient(endpoint string, apiVersionString string) (dockerclient, error) {
+	return godocker.NewVersionedClient(endpoint, apiVersionString)
+}
+
+func newDockerClient(dockerClientFactory dockerClientFactory, pingBackoff backoff.Backoff) (dockerclient, error) {
 	dockerUnixSocketSourcePath, fromEnv := config.DockerUnixSocket()
 	if !fromEnv {
 		dockerUnixSocketSourcePath = "/var/run/docker.sock"
 	}
-	client, err := godocker.NewVersionedClient(config.UnixSocketPrefix+dockerUnixSocketSourcePath, "1.15")
+	client, err := dockerClientFactory.NewVersionedClient(
+		config.UnixSocketPrefix+dockerUnixSocketSourcePath, "1.15")
 	if err != nil {
 		return nil, err
 	}
-	err = client.Ping()
+	for {
+		err = client.Ping()
+		if err != nil {
```

Comment:
I think this would be clearer like this:

```go
if err == nil {
	break
}
shouldRetry := (isNetworkError(err) || isRetryablePingError(err)) && pingBackoff.ShouldRetry()
if !shouldRetry {
	break
}
backoffDuration := pingBackoff.Duration()
log.Infof( /*...*/)
time.Sleep(backoffDuration)
```

That gets rid of the multiple levels of indentation and I think reads more clearly.